lib/testutils/src/proptest.rs

```diff
@@ -14,309 +14,249 @@
 
 use std::collections::BTreeMap;
 use std::fmt::Debug;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
 
-use itertools::Either;
 use itertools::Itertools as _;
-use jj_lib::repo_path::RepoPath;
+use jj_lib::backend::BackendResult;
+use jj_lib::backend::MergedTreeId;
+use jj_lib::backend::TreeValue;
+use jj_lib::merge::Merge;
+use jj_lib::merged_tree::MergedTreeBuilder;
 use jj_lib::repo_path::RepoPathBuf;
-use jj_lib::repo_path::RepoPathComponentBuf;
-use proptest::collection::btree_map;
-use proptest::collection::vec;
+use jj_lib::store::Store;
 use proptest::prelude::*;
-use proptest::sample::select;
+use proptest_derive::Arbitrary;
 use proptest_state_machine::ReferenceStateMachine;
 
-#[derive(Debug, Clone)]
-pub struct RepoRefState {
-    root: Tree,
-}
+use crate::write_file;
 
-#[derive(Debug, Clone)]
-pub enum File {
-    RegularFile { contents: String, executable: bool },
+fn arb_file_contents() -> impl Strategy<Value = String> {
+    prop_oneof![
+        Just("".to_string()),
+        // Diffing is line-oriented, so try to generate files with relatively
+        // many newlines.
+        "(\n|[a-z]|.)*".prop_map(|s| s.to_string()),
+    ]
 }
 
-#[derive(Debug, Clone)]
-pub enum Transition {
-    /// Create a file with the given contents at `path`.
-    ///
-    /// Parent directories are created as necessary, existing files or
-    /// directories are replaced.
-    CreateFile {
-        path: RepoPathBuf,
+#[derive(Arbitrary, Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum DirEntry {
+    File {
+        #[proptest(strategy = "arb_file_contents()")]
         contents: String,
```

Comment:
Considering that `arb_file_contents()` now can also generate "binary file" like `"\0"`, perhaps it would make sense to store the `contents` as a `Vec<u8>`? While the `.*` regex does generate "binary" data as interpreted by `buf_to_file_contents()` due to the inclusion of null terminators, `String` still enforces valid UTF-8.